lite2/client.go

```diff
@@ -708,72 +708,64 @@ func (c *Client) sequence(
+	newVals *types.ValidatorSet, // height g
 	now time.Time) error {
 
-	c.logger.Debug("Verify newHeader against lastHeader",
-		"lastHeight", lastHeader.Height,
-		"lastHash", lastHeader.Hash(),
-		"newHeight", newHeader.Height,
-		"newHash", newHeader.Hash())
-	err := Verify(c.chainID, lastHeader, lastVals, newHeader, newVals, c.trustingPeriod, now, c.trustLevel)
-	switch err.(type) {
-	case nil:
-		return nil
-	case ErrNewValSetCantBeTrusted:
-		// continue bisection
-	default:
-		return errors.Wrapf(err, "failed to verify the header #%d", newHeader.Height)
-	}
-
-	pivot := (c.trustedHeader.Height + newHeader.Header.Height) / 2
-	pivotHeader, pivotVals, err := c.fetchHeaderAndValsAtHeight(pivot)
-	if err != nil {
-		return err
-	}
-
-	// left branch
-	{
-		err := c.bisection(lastHeader, lastVals, pivotHeader, pivotVals, now)
-		if err != nil {
-			return errors.Wrapf(err, "bisection of #%d and #%d", lastHeader.Height, pivot)
-		}
-	}
+	interimVals := newVals
+	interimHeader := newHeader
 
-	// right branch
-	{
-		nextVals, err := c.validatorSetFromPrimary(pivot + 1)
-		if err != nil {
-			return errors.Wrapf(err, "failed to obtain the vals #%d", pivot+1)
-		}
-		if !bytes.Equal(pivotHeader.NextValidatorsHash, nextVals.Hash()) {
-			return errors.Errorf("expected next validator's hash %X, but got %X (height #%d)",
-				pivotHeader.NextValidatorsHash,
-				nextVals.Hash(),
-				pivot)
-		}
+	for trustedHeader.Height < newHeader.Height {
+		c.logger.Debug("Verify newHeader against trustedHeader",
+			"lastHeight", trustedHeader.Height,
+			"lastHash", trustedHeader.Hash(),
+			"newHeight", newHeader.Height,
+			"newHash", newHeader.Hash())
+		err := Verify(c.chainID, trustedHeader, trustedNextVals, interimHeader, interimVals, c.trustingPeriod, now,
+			c.trustLevel)
+		switch err.(type) {
+		case nil:
+
+			// Update the lower bound to the previous upper bound
+			trustedHeader = interimHeader
+			trustedNextVals, err = c.validatorSetFromPrimary(interimHeader.Height + 1)
+			if err != nil {
+				return err
+			}
+			if !bytes.Equal(trustedHeader.NextValidatorsHash, trustedNextVals.Hash()) {
+				return errors.Errorf("expected next validator's hash %X, but got %X (height #%d)",
+					trustedHeader.NextValidatorsHash,
+					trustedNextVals.Hash(),
+					trustedHeader.Height)
+			}
 
-		err = c.updateTrustedHeaderAndVals(pivotHeader, nextVals)
-		if err != nil {
-			return errors.Wrapf(err, "failed to update trusted state #%d", pivot)
-		}
+			err = c.updateTrustedHeaderAndNextVals(trustedHeader, trustedNextVals)
+			if err != nil {
+				return err
+			}
 
-		err = c.bisection(pivotHeader, nextVals, newHeader, newVals, now)
-		if err != nil {
-			return errors.Wrapf(err, "bisection of #%d and #%d", pivot, newHeader.Height)
+			// Update the upper bound to the untrustedHeader
+			interimHeader, interimVals = newHeader, newVals
+		case ErrNewValSetCantBeTrusted:
+			pivotHeight := (interimHeader.Height + trustedHeader.Height) / 2
```

Comment:
what if interimHeader.Height = 2 and trustedHeader.Height = 1

=> pivotHeight = 1 => Infinite loop?

1) should we use `math.Ceil` function
2) should we check that headers are adjacent and exit if header cannot be trusted? I think there should be another condition (`if interimHeader.Height == trustedHeader.Height+1`) besides trustedHeader.Height < newHeader.Height for bisection to exit.

Reply:
Yes I believe you are correct. I will need to change the code and write tests accordingly

Reply:
I actually think the changes to the other pr #4412 with the addition of the generalised `Verify` function will solve this

Reply:
Even if that is true, we still need to write more tests since current suite clearly did not cover ^ use-case.

Reply:
I will write some tests regarding the non recursive bisection. Can we first review and merge #4412 and then I will adjust the bisection method accordingly

Reply:
https://github.com/tendermint/tendermint/pull/4412#pullrequestreview-359887880 my review